File: apps/api/app/persistence/retention.py
"""Retention sweeping removes expired resume data across storage and tables.

Independent evaluations expire whole (source file, extracted facts, report,
improved DOCX). Employer documents expire with every submission, evaluation,
and embedding derived from them, so no candidate-identifying content survives
past the configured window.
"""

from dataclasses import dataclass
from datetime import datetime

from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from ..documents.storage import LocalObjectStorage
from .models import (
	BatchEvaluation,
	BatchEvaluationSubmission,
	Evaluation,
	IndependentEvaluation,
	ProcessingJob,
	ResumeDocument,
	ResumeSubmission,
	ResumeVersion,
)
from .points import release_in_session


@dataclass(frozen=True)
class RetentionResult:
	documents_purged: int
	independent_evaluations_purged: int


def independent_storage_keys(evaluation: IndependentEvaluation) -> list[str]:
	return [
		key
		for key in (
			evaluation.storage_key,
			evaluation.job_description_key,
			evaluation.improved_resume_key,
		)
		if key
	]


async def purge_expired_data(
	session: AsyncSession,
	storage: LocalObjectStorage,
	now: datetime,
) -> RetentionResult:
	documents = (
		await session.scalars(
			select(ResumeDocument)
			.where(ResumeDocument.retention_date <= now)
			.order_by(ResumeDocument.retention_date)
		)
	).all()
	evaluations = (
		await session.scalars(
			select(IndependentEvaluation)
			.where(IndependentEvaluation.retention_date <= now)
			.order_by(IndependentEvaluation.retention_date)
		)
	).all()

	document_keys = [document.storage_key for document in documents]
	evaluation_keys = [
		key
		for evaluation in evaluations
		for key in independent_storage_keys(evaluation)
	]

	await _purge_independent_evaluations(session, list(evaluations))
	await _purge_resume_documents(session, list(documents))

	# Objects go only after the deletes succeed; a crash between the two
	# leaves an unreachable object, never a dangling reference.
	for key in [*document_keys, *evaluation_keys]:
		storage.delete(key)

	return RetentionResult(
		documents_purged=len(documents),
		independent_evaluations_purged=len(evaluations),
	)


async def _purge_independent_evaluations(
	session: AsyncSession,
	evaluations: list[IndependentEvaluation],
) -> None:
	for evaluation in evaluations:
		if evaluation.point_reservation_id is not None:
			# An unsettled hold must return to the balance instead of
			# expiring with the report it paid for.
			await release_in_session(session, str(evaluation.point_reservation_id))
		await session.execute(
			delete(ProcessingJob).where(
				(ProcessingJob.type == "independent_evaluation_processing")
				& (ProcessingJob.payload_reference == evaluation.id)
			)
		)
		await session.delete(evaluation)


async def _purge_resume_documents(
	session: AsyncSession,
	documents: list[ResumeDocument],
) -> None:
	if not documents:
		return
	version_ids = select(ResumeVersion.id).where(
		ResumeVersion.resume_document_id.in_([document.id for document in documents])
	)
	submission_ids = select(ResumeSubmission.id).where(
		ResumeSubmission.resume_version_id.in_(version_ids)
	)
	# Evaluations first: their rows RESTRICT version deletion and their
	# children (assessments, review decisions) cascade from here.
	await session.execute(delete(Evaluation).where(Evaluation.resume_version_id.in_(version_ids)))
	await session.execute(
		delete(BatchEvaluationSubmission).where(
			BatchEvaluationSubmission.resume_submission_id.in_(submission_ids)
		)
	)
	await session.execute(delete(ResumeSubmission).where(ResumeSubmission.id.in_(submission_ids)))
	# Batches left without linked submissions or evaluations are empty shells.
	await session.execute(
		delete(BatchEvaluation).where(
			BatchEvaluation.id.not_in(select(BatchEvaluationSubmission.batch_evaluation_id))
			& BatchEvaluation.id.not_in(
				select(Evaluation.batch_evaluation_id).where(
					Evaluation.batch_evaluation_id.is_not(None)
				)
			)
		)
	)
	await session.execute(
		delete(ResumeDocument).where(ResumeDocument.id.in_([document.id for document in documents]))
	)

File: libs/server-core/src/retention.ts
import { sql } from "drizzle-orm";
import type { Database } from "./db.ts";
import { LocalObjectStorage } from "./storage.ts";
import { releaseInSession } from "./points.ts";

export type RetentionResult = {
	readonly documentsPurged: number;
	readonly independentEvaluationsPurged: number;
};

export const independentStorageKeys = (evaluation: { storageKey: string; jobDescriptionKey: string | null; improvedResumeKey: string | null }): string[] =>
	[evaluation.storageKey, evaluation.jobDescriptionKey, evaluation.improvedResumeKey].filter((key): key is string => Boolean(key));

export const purgeExpiredData = async (db: Database, storage: LocalObjectStorage, now: Date): Promise<RetentionResult> => {
	const documents = await db.execute(sql`SELECT id, storage_key FROM resume_document WHERE retention_date <= ${now} ORDER BY retention_date`);
	const evaluations = await db.execute(sql`SELECT id, storage_key, job_description_key, improved_resume_key, point_reservation_id FROM independent_evaluation WHERE retention_date <= ${now} ORDER BY retention_date`);
	const documentRows = documents as unknown as Array<{ id: string; storage_key: string }>;
	const evaluationRows = evaluations as unknown as Array<{ id: string; storage_key: string; job_description_key: string | null; improved_resume_key: string | null; point_reservation_id: string | null }>;

	for (const evaluation of evaluationRows) {
		if (evaluation.point_reservation_id) await releaseInSession(db, evaluation.point_reservation_id);
		await db.execute(sql`DELETE FROM processing_job WHERE type = 'independent_evaluation_processing' AND payload_reference = ${evaluation.id}`);
		await db.execute(sql`DELETE FROM independent_evaluation WHERE id = ${evaluation.id}`);
	}

	if (documentRows.length) {
		const ids = documentRows.map((item) => item.id);
		const versions = (await db.execute(sql`SELECT id FROM resume_version WHERE resume_document_id IN (${sql.join(ids.map((id) => sql`${id}`), sql`, `)})`)) as unknown as Array<{ id: string }>;
		const versionIds = versions.map((item) => item.id);
		if (versionIds.length) {
			const holds = (await db.execute(sql`SELECT point_reservation_id FROM evaluation WHERE resume_version_id IN (${sql.join(versionIds.map((id) => sql`${id}`), sql`, `)})`)) as unknown as Array<{ point_reservation_id: string | null }>;
			for (const hold of holds) if (hold.point_reservation_id) await releaseInSession(db, hold.point_reservation_id);
			await db.execute(sql`DELETE FROM evaluation WHERE resume_version_id IN (${sql.join(versionIds.map((id) => sql`${id}`), sql`, `)})`);
			const submissions = (await db.execute(sql`SELECT id FROM resume_submission WHERE resume_version_id IN (${sql.join(versionIds.map((id) => sql`${id}`), sql`, `)})`)) as unknown as Array<{ id: string }>;
			const submissionIds = submissions.map((item) => item.id);
			if (submissionIds.length) {
				await db.execute(sql`DELETE FROM batch_evaluation_submission WHERE resume_submission_id IN (${sql.join(submissionIds.map((id) => sql`${id}`), sql`, `)})`);
				await db.execute(sql`DELETE FROM resume_submission WHERE id IN (${sql.join(submissionIds.map((id) => sql`${id}`), sql`, `)})`);
			}
			await db.execute(sql`DELETE FROM batch_evaluation WHERE id NOT IN (SELECT batch_evaluation_id FROM batch_evaluation_submission) AND id NOT IN (SELECT batch_evaluation_id FROM evaluation WHERE batch_evaluation_id IS NOT NULL)`);
		}
		await db.execute(sql`DELETE FROM resume_document WHERE id IN (${sql.join(ids.map((id) => sql`${id}`), sql`, `)})`);
	}

	for (const key of [...documentRows.map((item) => item.storage_key), ...evaluationRows.flatMap((item) => independentStorageKeys({ storageKey: item.storage_key, jobDescriptionKey: item.job_description_key, improvedResumeKey: item.improved_resume_key }))]) {
		await storage.delete(key);
	}

	return { documentsPurged: documentRows.length, independentEvaluationsPurged: evaluationRows.length };
};
